src/client/fuse_ll.cc

```diff
@@ -1658,36 +1671,52 @@ uint64_t CephFuse::Handle::make_fake_ino(inodeno_t ino, snapid_t snapid)
     if (snapid == CEPH_NOSNAP && ino == client->get_root_ino())
       return FUSE_ROOT_ID;
 
-    std::lock_guard l(stag_lock);
-    auto p = snap_stag_map.find(snapid);
-    if (p != snap_stag_map.end()) {
-      inodeno_t fino = MAKE_FINO(ino, p->second);
-      return fino;
-    }
-
-    int first = last_stag & STAG_MASK;
-    int stag =  (++last_stag) & STAG_MASK;
-    for (; stag != first; stag = (++last_stag) & STAG_MASK) {
-      // two reserved stags: 0 for CPEH_NOSNAP and 1 for CEPH_SNAPDIR
-      if (stag == 0 || stag == 1)
-	continue;
-
-      auto p = stag_snap_map.find(stag);
-      if (p == stag_snap_map.end()) {
-	snap_stag_map[snapid] = stag;
-	stag_snap_map[stag] = snapid;
-	break;
+    int stag;
+    if (snapid == CEPH_NOSNAP) {
+      stag = G_NOSNAP_STAG;
+    } else if (snapid == CEPH_SNAPDIR) {
+      stag = G_SNAPDIR_STAG;
+    } else {
+      std::lock_guard l(stag_lock);
+      auto &fino_map = g_fino_map[ino]; // will insert it anyway if not exists
+      auto p = fino_map.snap_stag_map.find(snapid);
+      if (p != fino_map.snap_stag_map.end()) {
+        inodeno_t fino = MAKE_FINO(ino, p->second);
+        return fino;
```

Comment:
is possible, please make use of structured binding to annotate `p->second` attribute

Reply:
Done.